src/java.base/share/classes/java/io/BufferedInputStream.java

```diff
@@ -166,16 +171,42 @@ private InputStream getInIfOpen() throws IOException {
     }
 
     /**
-     * Check to make sure that buffer has not been nulled out due to
-     * close; if not return it;
+     * Returns the internal buffer, optionally allocating it if empty.
+     * @param allocateIfEmpty true to allocate if empty
+     * @throws IOException if the stream is closed (buf is null)
      */
-    private byte[] getBufIfOpen() throws IOException {
+    private byte[] getBufIfOpen(boolean allocateIfEmpty) throws IOException {
         byte[] buffer = buf;
-        if (buffer == null)
+        if (allocateIfEmpty && buffer == EMPTY) {
+            buffer = new byte[initialSize];
```

Comment:
It took me a few rounds of reading this code to realize that this array allocation will only be done once even when multiple threads are involved, because when `buffer == EMPTY` (i.e. `BufferedInputStream` is not subclassed), then the call to `getBufIfOpen(...)` always happens after acquiring the `lock`. So, only one thread will ever enter this block. Looks good to me.